docs/keyed_services.md

```diff
@@ -0,0 +1,153 @@
+    service->Bind(std::move(receiver));
+  }
+}
+
+static jlong JNI_FakeService_GetForProfile(
+    JNIEnv* env,
+    const base::android::JavaParamRef<jobject>& profile_android) {
+  auto* profile = Profile::FromJavaObject(profile_android);
+  auto pending =
+      FakeServiceFactory::GetInstance()->GetRemoteService(profile);
+  return pending.PassPipe().release().value()
+}
+```
+
+Java
+
+brave/browser/fake/android/java/src/org/chromium/brave/browser/fake/FakeServiceFactory.java
+```java
+public @Nullable FakeService getForProfile(Profile profile,
+        @Nullable ConnectionErrorHandler connectionErrorHandler) {
+    long nativeHandle = FakeServiceFactoryJni.get().getForProfile(profile);
+    MessagePipeHandle handle =
+            CoreImpl.getInstance().acquireNativeHandle(nativeHandle).
+            toMessagePipeHandle();
+    if (!handle.isValid()) {
+      return null;
+    }
+    FakeService fakeService =
+        FakeService.MANAGER.attachProxy(handle, 0);
+    if (connectionErrorHandler != null) {
+        Handler handler = ((Interface.Proxy) fakeService).getProxyHandler();
+        handler.setErrorHandler(connectionErrorHandler);
+    }
+    return fakeService;
+}
+```
+
+#### IOS
+
+IOS has separate factories because Profile subclasses `web::BrowserState` instead of `content::BrowserContext`, but the same principles apply to `ProfileKeyedServiceFactoryIOS`.
+
+#### IOS and Desktop profile checks
+
+TODO - standardize a way to maintain consistency between `GetBrowserContextToUse` methods on desktop and android. Possibly using some constants like `kRegular`, `kIncognito`, `kTor` so the same code can be used to check both `Profile` and `ProfileIOS`
+
+#### IOS keyed services and mojo
+
+obj-c
+
+brave/ios/browser/fake/fake_service_factory.mm
+```cpp
+mojo::PendingRemote<fake::mojom::FakeService> FakeServiceFactory::GetForProfile(
+    ProfileIOS* profile) {
+  auto* service = GetInstance()->GetServiceForProfileAs<fake::FakeServiceImpl>(
+      profile, true);
+  if (!service) {
+    return mojo::PendingRemote<fake::mojom::FakeService>();
+  }
+  return service->MakeRemote();
+}
+
+@implementation FakeFakeServiceFactory
++ (nullable id)getForProfile:(ProfileIOS*)profile {
+  auto* fake_service =
+      FakeServiceFactory::GetRemoteForProfile(profile);
+  if (!fake_service) {
+    return nil;
+  }
+  mojo::PendingRemote<fake::mojom::FakeService> pending_remote;
+  fake_service->Bind(pending_remote.InitWithNewPipeAndPassReceiver());
+  return [[FakeFakeServiceMojoImpl alloc]
+      initWithFakeService:std::move(pending_remote)];
+}
+```
+
+brave/ios/browser/fake/fake_service_factory_wrapper.h
+```objc
+#import <Foundation/Foundation.h>
+#include "keyed_service_factory_wrapper.h"  // NOLINT
+
+@protocol FakeFakeService;
+
+OBJC_EXPORT
+NS_SWIFT_NAME(Fake.FakeServiceFactory)
+@interface FakeFakeServiceFactory
+    : KeyedServiceFactoryWrapper < id <FakeFakeService>
+> @end
+
+#endif  // BRAVE_IOS_BROWSER_SKUS_SKUS_SDK_FACTORY_WRAPPERS_H_
```

Comment:
can probably remove this

Reply:
remove what? The entire wrapper?

Reply:
No sorry I just meant the line specifically, you have the end of the header guard in this example code

Reply:
oic, yep